### PR TITLE
LDAP: fix inGroup for memberUid type of group memberships

### LINKED ISSUE #24252: LDAP inGroup() failing
We can't move a directory into an other directory when the first one is a shared one, shared by group.

### Steps to reproduce
1. User A creates a directory DA, then share it to group G
2. User B (which is in group G) sees DA. Trying to drag&drop DA into an other (regular) directory failed, and in logs we get "InvalidArgumentException: Invalid recipient"

### Expected behaviour
Directory DA should be movable.
It is the case when directory is shared by user (not by group).
It is also the case when using local groups (not LDAP groups).

### Actual behaviour
Directory can't be moved into an other directory.

### Server configuration
Debian 10, Apache, MySQL (MariaDB), PHP7. All up-to-date installed from OS.

NextCloud 20.0.1, updated with OCC from 18.x. 

#### LDAP configuration
<details>
<summary>LDAP config</summary>

```
+-------------------------------+-----------------------------------+
| Configuration                 | s01                               |
+-------------------------------+-----------------------------------+
| hasMemberOfFilterSupport      | 0                                 |
| homeFolderNamingRule          | attr:uid                          |
| lastJpegPhotoLookup           | 0                                 |
| ldapAgentName                 | cn=l***   |
| ldapAgentPassword             | ***                               |
| ldapAttributesForGroupSearch  |                                   |
| ldapAttributesForUserSearch   |                                   |
| ldapBackupHost                | ***.***.***.***                   |
| ldapBackupPort                | 389                               |
| ldapBase                      | dc=***            |
| ldapBaseGroups                | ou=***  |
| ldapBaseUsers                 | ou=*** |
| ldapCacheTTL                  | 600                               |
| ldapConfigurationActive       | 1                                 |
| ldapDefaultPPolicyDN          |                                   |
| ldapDynamicGroupMemberURL     |                                   |
| ldapEmailAttribute            | mail                              |
| ldapExperiencedAdmin          | 0                                 |
| ldapExpertUUIDGroupAttr       |                                   |
| ldapExpertUUIDUserAttr        |                                   |
| ldapExpertUsernameAttr        | uid                               |
| ldapExtStorageHomeAttribute   |                                   |
| ldapGidNumber                 | gidNumber                         |
| ldapGroupDisplayName          | cn                                |
| ldapGroupFilter               | (objectclass=*)                   |
| ldapGroupFilterGroups         |                                   |
| ldapGroupFilterMode           | 1                                 |
| ldapGroupFilterObjectclass    |                                   |
| ldapGroupMemberAssocAttr      | memberUid                         |
| ldapHost                      | ***                   |
| ldapIgnoreNamingRules         |                                   |
| ldapLoginFilter               | uid=%uid                          |
| ldapLoginFilterAttributes     |                                   |
| ldapLoginFilterEmail          | 0                                 |
| ldapLoginFilterMode           | 1                                 |
| ldapLoginFilterUsername       | 1                                 |
| ldapMatchingRuleInChainState  | unknown                           |
| ldapNestedGroups              | 0                                 |
| ldapOverrideMainServer        |                                   |
| ldapPagingSize                | 500                               |
| ldapPort                      | 389                               |
| ldapQuotaAttribute            |                                   |
| ldapQuotaDefault              |                                   |
| ldapTLS                       | 1                                 |
| ldapUserAvatarRule            | default                           |
| ldapUserDisplayName           | displayname                       |
| ldapUserDisplayName2          |                                   |
| ldapUserFilter                | (objectclass=*)                   |
| ldapUserFilterGroups          |                                   |
| ldapUserFilterMode            | 1                                 |
| ldapUserFilterObjectclass     |                                   |
| ldapUuidGroupAttribute        | auto                              |
| ldapUuidUserAttribute         | auto                              |
| turnOffCertCheck              | 1                                 |
| turnOnPasswordChange          | 0                                 |
| useMemberOfToDetectMembership | 1                                 |
+-------------------------------+-----------------------------------+
```
</details>

#### Debug / workaround

I tracked down the error "InvalidArgumentException: Invalid recipient" and found that the method `inGroup()` in file `apps/user_ldap/lib/Group_LDAP.php` replies FALSE when I try to move directory DA into a local directory. I am in group G, and DA is shared by its owner to LDAP group G (note: all users come from LDAP).

After added some debug, I found that `$members` table holds the LDAP entries for the tested group (here group G).
The `inGroup()` method uses `in_array($userDN, $members)` to check if user is in the group.
But : this table is not a flat table. It is an array with one line per member, each line is an array of `property => value` with properties stuff like "mail", "dn", "displayName"... and value an array containing the value.
So the actual DN for a given user is stored in `$members[index]['dn'][0]`.

I replaced the `in_array` with a loop: `foreach($members as $m) { if (in_array($userDN, $m['dn']) { $isInGroup=TRUE; beak; } }` and the problem is solved (after cleaning caches, and I made this change at the two places were `in_array` is used).

Is it a problem with our LDAP configuration or a bug in the `inGroup()`? Regards.

### PR DESCRIPTION
**Pull request by blizzz**
fixes #24252

basically, the code that resolves the rdns to full DNs returned the whole records, as @Hexasoft observed, but the following code requires a flat list of DNs. 

@Hexasoft @steffen-moser can you confirm it fixes the issue for you?

@tofuSCHNITZEL you are informed because it is relevant for _zimbramailforwardingaddress_ types, I would be surprised if it breaks anything though.

It set the label to "developing", because i want to check one, two things tomorrow.

### CODE REVIEW

**Comment by steffen-moser** [comment]
@blizzz - I think you also need to consider https://github.com/nextcloud/server/blob/57bfe0d1f9a7dc5ad545a156c17ada1353fe6c94/apps/user_ldap/lib/Group_LDAP.php#L112 as the list there is also not flat according to my debug log.

I've not checked whether https://github.com/nextcloud/server/blob/57bfe0d1f9a7dc5ad545a156c17ada1353fe6c94/apps/user_ldap/lib/Group_LDAP.php#L119 needs fixing because I did not run into this if clause on any of our Nextcloud servers.

**Comment by tofuSCHNITZEL** [comment]
I just ran a few tests and you change makes total sense... now _$members_ really contains dns so the in_array check works.
But there are problems if this is tripped
https://github.com/nextcloud/server/blob/57bfe0d1f9a7dc5ad545a156c17ada1353fe6c94/apps/user_ldap/lib/Group_LDAP.php#L160
because then $dns is already filled but ins this form:

```
Array
(
    [0] => Array
        (
            [entryuuid] => Array
                (
                    [0] => c2a1b704-4463-1038-9b2b-b7afed259854
                )

            [dn] => Array
                (
                    [0] => uid=testaccount,ou=people,dc=email,dc=at
                )

            [uid] => Array
                (
                    [0] => testaccount
                )

            [displayname] => Array
                (
                    [0] => testaccount
                )

        )

    [1] => Array
        (
            [entryuuid] => Array
                (
                    [0] => 77dd1308-5857-1030-8796-392f61712f0b
                )

            [dn] => Array
                (
                    [0] => uid=user.name,ou=people,dc=email,dc=at
                )

            [uid] => Array
                (
                    [0] => user.name
                )

            [zimbramailalias] => Array
                (
                    [0] => user@email.at
                )

            [displayname] => Array
                (
                    [0] => user
                )

        )

)
```

and now the array_reduce part might or might not be run (depending if there are filterParts left)
if its run then using $dns as initial with the array_reduce does not produces the desired output.
if its not run then we have nonsense in _$members_ like before this fix

TLDR:
after over an hour of playing around with different scenarios I think I found the solution that works with all cases.

I committed my changes to this branch... I hope this is okay? (and the proper way?)

I'm also thinking, that a foreach for cleaning up the array would be better to understand than the array_reduce.

we could replace:
```
				// now we cleanup the users array to get only dns
				$dns = array_reduce($users, function (array $carry, array $record) {
					if (!in_array($carry, $record['dn'][0])) {
						$carry[$record['dn'][0]] = 1;
					}
					return $carry;
				}, []);
				$members = array_keys($dns);
```
with a simple:

				// now create a list of dns
				$members = [];
				foreach ($users as $user){
					$dn = $user['dn'][0];
					if (!in_array($dn, $members))
						$members[] = $dn;
				}


also should we could (should?) move:
https://github.com/nextcloud/server/blob/57bfe0d1f9a7dc5ad545a156c17ada1353fe6c94/apps/user_ldap/lib/Group_LDAP.php#L139-L141
after the switch statement

**Comment by blizzz** [comment]
>  as the list there is also not flat according to my debug log.

is it possible that this was old information, perhaps pulled it from caching? Since it really only is assigned the value of `$members`. Also not that cachedGroupMembers is not a plain array, but an instance of `CappedMemoryCache`  that behaves like one.

**Comment by blizzz** [comment]
@tofuSCHNITZEL good catch with the loop! Also thanks for the commit, that's appreciated. I'd build upon it and turn the last array_reduce to an unsexy, but less hungry and more performant foreach (like you suggest). I'd still work with having the DN as index, as it saves in_arrays and prevents potential duplicates.

Moving the count check down makes sense as well. And dropping the is_array check with it since it is guaranteed by the return type.

**Comment by tofuSCHNITZEL** [comment]
great ;) should / will this fix be backported to 19 or even earlier? I guess sind its not a huge change we should/could do it. the structure of the file is a bit different pre v20 because of my changes to it for the zimbra support so I could port the changes to the v19 GroupLDAP file.

**Comment by blizzz** [comment]
…and added unit tests.

> great ;) should / will this fix be backported to 19 or even earlier? I guess sind its not a huge change we should/could do it. the structure of the file is a bit different pre v20 because of my changes to it for the zimbra support so I could port the changes to the v19 GroupLDAP file.

Yes, it is a bug fix, and should be brought back down to 18. If we need to touch a bit there, so be it  :)

**Comment by steffen-moser** [comment]
> > as the list there is also not flat according to my debug log.
> 
> is it possible that this was old information, perhaps pulled it from caching? Since it really only is assigned the value of `$members`. Also not that cachedGroupMembers is not a plain array, but an instance of `CappedMemoryCache` that behaves like one.

Of course, It is very well possible that this old information came from caching.

We encountered the problem on two servers, but only one of them is actually running using a cache. And only on this server, the if clause, where line 112 resides, was entered, and only there I needed the fix. I haven't done further testing, but maybe emptying the cache would have also helped.

**Comment by Hexasoft** [comment]
Sorry, I was too busy at work to test the last proposal.
Thank you for your job. For the last change, not sure if it can occurs, but is it certain that `$dn = $user['dn'][0];` is always defined? I mean, does adding `if (isset($user['dn'][0])) {}` is of any interest?

**Comment by tofuSCHNITZEL** [comment]
> Thank you for your job. For the last change, not sure if it can occurs, but is it certain that `$dn = $user['dn'][0];` is always defined? I mean, does adding `if (isset($user['dn'][0])) {}` is of any interest?

I think, since dn is such an integral attribute to ldap, it will always be a part of a ldap query response.

**Comment by blizzz** [comment]
> > Thank you for your job. For the last change, not sure if it can occurs, but is it certain that `$dn = $user['dn'][0];` is always defined? I mean, does adding `if (isset($user['dn'][0])) {}` is of any interest?
> 
> I think, since dn is such an integral attribute to ldap, it will always be a part of a ldap query response.

Yes, it always returns the dn as it is the de-facto (current!) identifier, even if you ask for a different attribute:

```
dapsearch -LLL -a find -H ldap://localhost:7770 -D cn=root,dc=nc,dc=zara -W -b "ou=small,dc=nc,dc=zara" "uid=*" uid  
Enter LDAP Password: 
dn: uid=folk_0,ou=users,ou=small,dc=nc,dc=zara
uid: folk_0

dn: uid=folk_1,ou=users,ou=small,dc=nc,dc=zara
uid: folk_1

dn: uid=folk_2,ou=users,ou=small,dc=nc,dc=zara
uid: folk_2
…
```

**Review by ChristophWurst — APPROVED** [review]
:+1: :elephant:

**Review by tofuSCHNITZEL — APPROVED** [review]
lgtm

**Comment by tofuSCHNITZEL** [comment]
@blizzz how will we do the stable19 backport? will we create a new branch and do the changes there manually? on what commit will this branch be based?

**Comment by blizzz** [comment]
> @blizzz how will we do the stable19 backport? will we create a new branch and do the changes there manually? on what commit will this branch be based?

Yes, we should do it. It's somewhere deep in my list, but if you want to beat me, that would be really appreciated :) A branch called `backport/24402/stable19` shall be based against stable19. 'git cherry-pick' the commits in chronological order, and `git mergetool` to resolve conflicts.

**Comment by tofuSCHNITZEL** [comment]
> 
> 
> > @blizzz how will we do the stable19 backport? will we create a new branch and do the changes there manually? on what commit will this branch be based?
> 
> Yes, we should do it. It's somewhere deep in my list, but if you want to beat me, that would be really appreciated :) A branch called `backport/24402/stable19` shall be based against stable19. 'git cherry-pick' the commits in chronological order, and `git mergetool` to resolve conflicts.

okay tried my best ;) hopefully I did everything how its supposed to be done ;)
it was interesting that the auto merging at the first cherry-pick replaced the wrong `$dns = array_merge($dns, $users);` (the one inside the over 9000 if) I guess it just matched the first "similar" line occurrence since the line numbers were different

**Comment by steffen-moser** [comment]
Are there any plans of backporting this to Nextcloud 19?
